nmigen/build/generic_platform.py

```diff
@@ -0,0 +1,161 @@
+from collections import defaultdict, OrderedDict
+
+from .. import *
+from ..hdl.rec import *
+from ..lib.io import *
+
+from .dsl import *
+from .tools import *
+
+
+__all__ = ["ConstraintError", "ConstraintManager"]
+
+
+class ConstraintError(Exception):
+    pass
+
+
+class ConstraintManager:
+    def __init__(self, resources):
+        self.available    = defaultdict(dict)
+        self.matched      = OrderedDict()
+        self.clocks       = OrderedDict()
+        self.tristates    = []
+        self.diffpairs    = []
+        self.ports        = []
+        self.inouts       = []
```

Comment:
Also, since none of the fields above appear to be a part of the public API, they all should be named starting with an underscore.